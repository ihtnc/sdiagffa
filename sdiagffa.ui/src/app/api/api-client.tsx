const baseUrl = 'https://localhost:44300';

export interface Person {
  id: number,
  name: string,
  gender: string,
  birthYear: string,
  height: string,
  mass: string,
  skinColor: string,
  eyeColor: string
};

export interface Film {
  id: number,
  title: string,
  releaseDate: string,
  episodeId: number,
  director: string,
  producer: string
};

export interface Planet {
  id: number,
  name: string,
  terrain: string,
  climate: string,
  surfaceWater: string,
  gravity: string,
  population: string,
  diameter: string
};

export interface Connection {
  personId: number,
  name: string,
  connection: 'Film' | 'Homeworld' | 'None',
  connectionId: number | null,
  connectionName: string | null,
  next: Connection | null
};

export interface PagedResults<T> {
  count: number,
  nextPageUrl: string | null,
  previousPageUrl: string | null,
  results: Array<T>
};

export const searchPeople = async function (search: string, page: number | null = null)
  : Promise<PagedResults<Person>> {
  const findUrl = '/details/people';

  const queries = [];
  if (search.length > 0) { queries.push(`search=${search}`); }
  if (page !== null) { queries.push(`page=${page}`); }
  const queryString = queries.length > 0 ? `?${queries.join('&')}` : '';

  const data = await fetch(`${baseUrl}${findUrl}${queryString}`);
  return data.json() as unknown as PagedResults<Person>;
};

export const customPeopleSearch = async function (peopleSearchUrl: string)
  : Promise<PagedResults<Person>> {
  const data = await fetch(`${baseUrl}${peopleSearchUrl}`);
  return data.json() as unknown as PagedResults<Person>;
};

export const findConnection = async function (referencePersonId: number, targetPersonId: number)
  : Promise<Connection> {
    const findConnectionUrl = '/find-connection';
    const requestOptions = {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        referencePersonId,
        targetPersonId
      })
    };
    const data = await fetch(`${baseUrl}${findConnectionUrl}`, requestOptions);
    return data.json() as unknown as Connection;
};

export const getPerson = async function (id: number)
  : Promise<Person> {
  const getUrl = '/details/people';

  const data = await fetch(`${baseUrl}${getUrl}/${id}`);
  return data.json() as unknown as Person;
};


export const getFilm = async function (id: number)
  : Promise<Film> {
  const getUrl = '/details/films';

  const data = await fetch(`${baseUrl}${getUrl}/${id}`);
  return data.json() as unknown as Film;
};

export const getPlanet = async function (id: number)
  : Promise<Planet> {
  const getUrl = '/details/planets';

  const data = await fetch(`${baseUrl}${getUrl}/${id}`);
  return data.json() as unknown as Planet;
};