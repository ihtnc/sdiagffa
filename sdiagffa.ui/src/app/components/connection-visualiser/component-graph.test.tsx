import { beforeEach, afterEach, describe, expect, it, vi } from "vitest";
import { render } from '@testing-library/react'
import ConnectionGraph from "./connection-graph";
import { Connection } from "@/api/api-client";
import { ConnectionId } from ".";

vi.mock('react-graph-vis', () => {
  const mock = vi.fn().mockImplementation(({ graph, options, events }) => {
    const nodes = [{
      id: 'Person_1',
      label: 'Luke Skywalker'
    }];
    events.select({nodes});

    return (<>
      <div>Graph: {JSON.stringify(graph)}</div>
      <div>Options: {JSON.stringify(options)}</div>
    </>);
  });
  return { default: mock };
});

vi.mock('usehooks-ts', () => {
  const useDarkMode = vi.fn();
  return { useDarkMode };
});

describe('ConnectionGraph component', () => {
  let useDarkModeMock;

  beforeEach(async () => {
    useDarkModeMock = vi.fn();
    const useHooks = await import('usehooks-ts');
    useHooks.useDarkMode = useDarkModeMock;

    useDarkModeMock.mockImplementation(() => {
      return { isDarkMode: false }
    });
  });

  afterEach(() => {
    useDarkModeMock.mockReset();
  });

  it('should render default state', async () => {
    const connection: Connection = {
      personId: 1,
      name: 'Luke Skywalker',
      connectionId: 0,
      connection: "None",
      connectionName: null,
      next: null
    };
    const component = render(<ConnectionGraph connection={connection} onSelect={vi.fn()} />);
    expect(component).toMatchSnapshot();
  });

  it('should render default state in darkmode', async () => {
    useDarkModeMock.mockImplementation(() => {
      return { isDarkMode: true }
    });

    const connection: Connection = {
      personId: 1,
      name: 'Luke Skywalker',
      connectionId: 0,
      connection: "None",
      connectionName: null,
      next: null
    };
    const component = render(<ConnectionGraph connection={connection} onSelect={vi.fn()} />);
    expect(component).toMatchSnapshot();
  });

  it('should call onSelect', async () => {
    const onSelect = vi.fn<[id: ConnectionId], void>();

    const connection: Connection = {
      personId: 1,
      name: 'Luke Skywalker',
      connectionId: 0,
      connection: "None",
      connectionName: null,
      next: null
    };
    render(<ConnectionGraph connection={connection} onSelect={onSelect} />);

    expect(onSelect).toHaveBeenCalled();
  });
});