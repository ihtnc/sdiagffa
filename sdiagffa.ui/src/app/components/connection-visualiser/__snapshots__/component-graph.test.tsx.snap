// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConnectionGraph component > should render default state 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="w-full h-full absolute top-0 left-0"
      >
        <div>
          Graph: 
          {"nodes":[{"id":"Character_1","label":"Luke Skywalker"}],"edges":[]}
        </div>
        <div>
          Options: 
          {"nodes":{"shape":"icon","icon":{"face":"FontAwesome","code":"","color":"#1B1614","weight":"normal"},"font":{"color":"#1B1614"},"chosen":{},"fixed":{"x":false,"y":false}},"edges":{"color":"#1B1614","font":{"color":"#1B1614","strokeWidth":0},"chosen":{},"width":2},"physics":{"enabled":false},"interaction":{"selectConnectedEdges":false}}
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="w-full h-full absolute top-0 left-0"
    >
      <div>
        Graph: 
        {"nodes":[{"id":"Character_1","label":"Luke Skywalker"}],"edges":[]}
      </div>
      <div>
        Options: 
        {"nodes":{"shape":"icon","icon":{"face":"FontAwesome","code":"","color":"#1B1614","weight":"normal"},"font":{"color":"#1B1614"},"chosen":{},"fixed":{"x":false,"y":false}},"edges":{"color":"#1B1614","font":{"color":"#1B1614","strokeWidth":0},"chosen":{},"width":2},"physics":{"enabled":false},"interaction":{"selectConnectedEdges":false}}
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`ConnectionGraph component > should render default state in darkmode 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="w-full h-full absolute top-0 left-0"
      >
        <div>
          Graph: 
          {"nodes":[{"id":"Character_1","label":"Luke Skywalker"}],"edges":[]}
        </div>
        <div>
          Options: 
          {"nodes":{"shape":"icon","icon":{"face":"FontAwesome","code":"","color":"#FFFFFF","weight":"normal"},"font":{"color":"#FFFFFF"},"chosen":{},"fixed":{"x":false,"y":false}},"edges":{"color":"#FFFFFF","font":{"color":"#FFFFFF","strokeWidth":0},"chosen":{},"width":2},"physics":{"enabled":false},"interaction":{"selectConnectedEdges":false}}
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="w-full h-full absolute top-0 left-0"
    >
      <div>
        Graph: 
        {"nodes":[{"id":"Character_1","label":"Luke Skywalker"}],"edges":[]}
      </div>
      <div>
        Options: 
        {"nodes":{"shape":"icon","icon":{"face":"FontAwesome","code":"","color":"#FFFFFF","weight":"normal"},"font":{"color":"#FFFFFF"},"chosen":{},"fixed":{"x":false,"y":false}},"edges":{"color":"#FFFFFF","font":{"color":"#FFFFFF","strokeWidth":0},"chosen":{},"width":2},"physics":{"enabled":false},"interaction":{"selectConnectedEdges":false}}
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
